pkg/plugins/resources/cargopackage/main.go

```diff
@@ -0,0 +1,198 @@
+	return cp.foundVersion.GetVersion(), versions, nil
+}
+
+func getPackageFileDir(packageName string) (string, error) {
+	if packageName == "" {
+		err := errors.New("got empty package name")
+		logrus.Errorf("%q\n", err)
+		return "", err
+	}
+	switch packageNameLen := len(packageName); packageNameLen {
+	case 1:
+		return fmt.Sprintf("%d", packageNameLen), nil
+	case 2:
+		return fmt.Sprintf("%d", packageNameLen), nil
+	case 3:
+		return fmt.Sprintf("%d/%s", packageNameLen, string(packageName[0])), nil
+	default:
+		return fmt.Sprintf("%s/%s", packageName[0:2], packageName[2:4]), nil
+	}
+}
+
+// Get package data from Json API
+func (cp *CargoPackage) getPackageData() (PackageData, error) {
+	var pd PackageData
+
+	pd.Name = cp.spec.Package
+
+	packageDir, err := getPackageFileDir(cp.spec.Package)
+	if err != nil {
+		logrus.Errorf("something went wrong while getting the package directory from its name %q\n", err)
+		return pd, err
+	}
+	packageFilePath := filepath.Join(cp.spec.IndexDir, packageDir, cp.spec.Package)
+	packageInfoFile, err := os.Open(packageFilePath)
+	if err != nil {
+		logrus.Errorf("something went wrong while opening the package file %q\n", err)
+		return pd, err
+	}
+	defer func(packageInfoFile *os.File) {
+		err := packageInfoFile.Close()
+		if err != nil {
+			logrus.Errorf("something went wrong while cleaning the package file %q\n", err)
+		}
+	}(packageInfoFile)
+
+	scanner := bufio.NewScanner(packageInfoFile)
+	for scanner.Scan() {
+		var packageVersion PackageVersion
+		err = json.Unmarshal(scanner.Bytes(), &packageVersion)
+		if err != nil {
+			logrus.Errorf("something went wrong while parsing the version %q\n", err)
+		}
+		if packageVersion.Yanked {
+			continue
+		}
+		pd.Versions = append(pd.Versions, packageVersion)
+	}
+	return pd, nil
+}
+
+func loadDefaultCrateIndex() (string, error) {
+	err := tmp.Create()
+	if err != nil {
+		return "", err
+	}
+	directory := tmp.Directory
+	indexPath := filepath.Join(directory, "crates-default-index")
+	if _, err := os.Stat(indexPath); os.IsNotExist(err) {
+		_, err = git.PlainClone(indexPath, false, &git.CloneOptions{
+			URL: cratesDefaultIndex,
+			// Can be reanabled once https://github.com/go-git/go-git/issues/305 is fixed
+			//Depth:        1,
+			SingleBranch: true,
+		})
+		if err != nil {
+			return "", err
+		}
+	} else {
+		// Repo already exists, update to latest
+		rep, err := git.PlainOpen(indexPath)
+		if err != nil {
+			logrus.Errorf("something went wrong when opening the existing repo %q\n", err)
+			return "", err
+		}
+		w, err := rep.Worktree()
+		if err != nil {
+			logrus.Errorf("something went wrong when loading the worktree of the existing repo %q\n", err)
+			return "", err
+		}
+		err = w.Pull(&git.PullOptions{})
+		if err != nil &&
+			err != git.NoErrAlreadyUpToDate {
+			logrus.Errorf("something went wrong when pulling the existing repo %q\n", err)
+			return "", err
+		}
+	}
+	if err != nil {
+		return "", err
+	}
+	return indexPath, nil
+}
```

Comment:
Is there a way to inject a `scm` entry from a condition or source context in order to prevent reimplementing `git clone` as it's a feature handled by `scm`?